diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/i915_drv.c archlinux-linux.new/drivers/gpu/drm/i915/i915_drv.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/i915_drv.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/i915_drv.c	2018-09-13 07:43:57.437676520 -0600
@@ -2691,7 +2691,7 @@
 	struct drm_i915_private *dev_priv = to_i915(dev);
 	int ret;
 
-	if (WARN_ON_ONCE(!(dev_priv->gt_pm.rc6.enabled && HAS_RC6(dev_priv))))
+	if (WARN_ON_ONCE(!(dev_priv->gt_pm.rc6.enabled && intel_rc6_enabled())))
 		return -ENODEV;
 
 	if (WARN_ON_ONCE(!HAS_RUNTIME_PM(dev_priv)))

diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/i915_drv.h archlinux-linux.new/drivers/gpu/drm/i915/i915_drv.h
--- archlinux-linux.orig/drivers/gpu/drm/i915/i915_drv.h	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/i915_drv.h	2018-09-13 07:43:57.437676520 -0600
@@ -2515,7 +2515,6 @@
 
 #define HAS_RC6(dev_priv)		 (INTEL_INFO(dev_priv)->has_rc6)
 #define HAS_RC6p(dev_priv)		 (INTEL_INFO(dev_priv)->has_rc6p)
-#define HAS_RC6pp(dev_priv)		 (false) /* HW was never validated */
 
 #define HAS_CSR(dev_priv)	(INTEL_INFO(dev_priv)->display.has_csr)
 
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/i915_params.c archlinux-linux.new/drivers/gpu/drm/i915/i915_params.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/i915_params.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/i915_params.c	2018-09-13 07:43:57.437676520 -0600
@@ -44,6 +44,13 @@
 	"Use kernel modesetting [KMS] (0=disable, "
 	"1=on, -1=force vga console preference [default])");
 
+i915_param_named_unsafe(enable_rc6, int, 0400,
+	"Enable power-saving render C-state 6. "
+	"Different stages can be selected via bitmask values "
+	"(0 = disable; 1 = enable rc6; 2 = enable deep rc6; 4 = enable deepest rc6). "
+	"For example, 3 would enable rc6 and deep rc6, and 7 would enable everything. "
+	"default: -1 (use per-chip default)");
+
 i915_param_named_unsafe(enable_dc, int, 0400,
 	"Enable power-saving display C-states. "
 	"(-1=auto [default]; 0=disable; 1=up to DC5; 2=up to DC6)");
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/i915_params.h archlinux-linux.new/drivers/gpu/drm/i915/i915_params.h
--- archlinux-linux.orig/drivers/gpu/drm/i915/i915_params.h	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/i915_params.h	2018-09-13 07:43:57.437676520 -0600
@@ -48,6 +48,7 @@
 	param(int, lvds_channel_mode, 0) \
 	param(int, panel_use_ssc, -1) \
 	param(int, vbt_sdvo_panel_type, -1) \
+	param(int, enable_rc6, -1) \
 	param(int, enable_dc, -1) \
 	param(int, enable_fbc, -1) \
 	param(int, enable_psr, -1) \
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/i915_pmu.c archlinux-linux.new/drivers/gpu/drm/i915/i915_pmu.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/i915_pmu.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/i915_pmu.c	2018-09-13 11:46:59.797382316 -0600
@@ -435,12 +435,13 @@
 				     VLV_GT_RENDER_RC6 :
 				     GEN6_GT_GFX_RC6);
 
-	if (HAS_RC6p(i915))
-		val += intel_rc6_residency_ns(i915, GEN6_GT_GFX_RC6p);
-
-	if (HAS_RC6pp(i915))
-		val += intel_rc6_residency_ns(i915, GEN6_GT_GFX_RC6pp);
+	if (HAS_RC6p(i915)) {
+		val += intel_rc6_residency_ns(i915,
+					      GEN6_GT_GFX_RC6p);
 
+		val += intel_rc6_residency_ns(i915,
+					      GEN6_GT_GFX_RC6pp);
+	}
 	return val;
 }
 
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/i915_sysfs.c archlinux-linux.new/drivers/gpu/drm/i915/i915_sysfs.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/i915_sysfs.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/i915_sysfs.c	2018-09-13 07:43:57.437676520 -0600
@@ -54,18 +54,7 @@
 static ssize_t
 show_rc6_mask(struct device *kdev, struct device_attribute *attr, char *buf)
 {
-	struct drm_i915_private *dev_priv = kdev_minor_to_i915(kdev);
-	unsigned int mask;
-
-	mask = 0;
-	if (HAS_RC6(dev_priv))
-		mask |= BIT(0);
-	if (HAS_RC6p(dev_priv))
-		mask |= BIT(1);
-	if (HAS_RC6pp(dev_priv))
-		mask |= BIT(2);
-
-	return snprintf(buf, PAGE_SIZE, "%x\n", mask);
+	return snprintf(buf, PAGE_SIZE, "%x\n", intel_rc6_enabled());
 }
 
 static ssize_t
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/intel_drv.h archlinux-linux.new/drivers/gpu/drm/i915/intel_drv.h
--- archlinux-linux.orig/drivers/gpu/drm/i915/intel_drv.h	2019-07-12 14:06:33.692489198 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/intel_drv.h	2019-07-12 14:06:04.078946949 -0600
@@ -1821,6 +1821,13 @@
 void intel_dp_start_link_train(struct intel_dp *intel_dp);
 void intel_dp_stop_link_train(struct intel_dp *intel_dp);
 
+/* re-enable rc6 support */
+int sanitize_rc6_option(struct drm_i915_private *dev_priv, int enable_rc6);
+static inline int intel_rc6_enabled(void)
+{
+       return i915_modparams.enable_rc6;
+}
+  
 /* intel_vdsc.c */
 int intel_dp_compute_dsc_params(struct intel_dp *intel_dp,
 				struct intel_crtc_state *pipe_config);
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/intel_guc.c archlinux-linux.new/drivers/gpu/drm/i915/intel_guc.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/intel_guc.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/intel_guc.c	2018-09-13 07:43:57.437676520 -0600
@@ -491,8 +491,9 @@
 	u32 action[2];
 
 	action[0] = INTEL_GUC_ACTION_SAMPLE_FORCEWAKE;
-	/* WaRsDisableCoarsePowerGating:skl,cnl */
-	if (!HAS_RC6(dev_priv) || NEEDS_WaRsDisableCoarsePowerGating(dev_priv))
+	/* WaRsDisableCoarsePowerGating:skl,bxt */
+	if (!intel_rc6_enabled() ||
+	    NEEDS_WaRsDisableCoarsePowerGating(dev_priv))
 		action[1] = 0;
 	else
 		/* bit 0 and 1 are for Render and Media domain separately */
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/intel_pm.c archlinux-linux.new/drivers/gpu/drm/i915/intel_pm.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/intel_pm.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/intel_pm.c	2018-09-13 07:43:57.447676586 -0600
@@ -56,6 +56,9 @@
  * which brings the most power savings; deeper states save more power, but
  * require higher latency to switch to and wake up.
  */
+#define INTEL_RC6_ENABLE			(1<<0)
+#define INTEL_RC6p_ENABLE			(1<<1)
+#define INTEL_RC6pp_ENABLE			(1<<2)
 
 static void gen9_init_clock_gating(struct drm_i915_private *dev_priv)
 {
@@ -6794,6 +6797,26 @@
 	I915_WRITE(GEN6_RP_CONTROL, 0);
 }
 
+static void intel_print_rc6_info(struct drm_i915_private *dev_priv, u32 mode)
+{
+	if (IS_VALLEYVIEW(dev_priv) || IS_CHERRYVIEW(dev_priv)) {
+		if (mode & (GEN7_RC_CTL_TO_MODE | GEN6_RC_CTL_EI_MODE(1)))
+			mode = GEN6_RC_CTL_RC6_ENABLE;
+		else
+			mode = 0;
+	}
+	if (HAS_RC6p(dev_priv))
+		DRM_DEBUG_DRIVER("Enabling RC6 states: "
+				 "RC6 %s RC6p %s RC6pp %s\n",
+				 onoff(mode & GEN6_RC_CTL_RC6_ENABLE),
+				 onoff(mode & GEN6_RC_CTL_RC6p_ENABLE),
+				 onoff(mode & GEN6_RC_CTL_RC6pp_ENABLE));
+
+	else
+		DRM_DEBUG_DRIVER("Enabling RC6 states: RC6 %s\n",
+				 onoff(mode & GEN6_RC_CTL_RC6_ENABLE));
+}
+
 static bool bxt_check_bios_rc6_setup(struct drm_i915_private *dev_priv)
 {
 	bool enable_rc6 = true;
@@ -6854,30 +6877,42 @@
 	return enable_rc6;
 }
 
-static bool sanitize_rc6(struct drm_i915_private *i915)
+int sanitize_rc6_option(struct drm_i915_private *dev_priv, int enable_rc6)
 {
-	struct intel_device_info *info = mkwrite_device_info(i915);
+	/* No RC6 before Ironlake and code is gone for ilk. */
+	if (INTEL_INFO(dev_priv)->gen < 6)
+		return 0;
 
-	/* Powersaving is controlled by the host when inside a VM */
-	if (intel_vgpu_active(i915))
-		info->has_rc6 = 0;
+	if (!enable_rc6)
+		return 0;
 
-	if (info->has_rc6 &&
-	    IS_GEN9_LP(i915) && !bxt_check_bios_rc6_setup(i915)) {
+	if (IS_GEN9_LP(dev_priv) && !bxt_check_bios_rc6_setup(dev_priv)) {
 		DRM_INFO("RC6 disabled by BIOS\n");
-		info->has_rc6 = 0;
+		return 0;
 	}
 
-	/*
-	 * We assume that we do not have any deep rc6 levels if we don't have
-	 * have the previous rc6 level supported, i.e. we use HAS_RC6()
-	 * as the initial coarse check for rc6 in general, moving on to
-	 * progressively finer/deeper levels.
-	 */
-	if (!info->has_rc6 && info->has_rc6p)
-		info->has_rc6p = 0;
+	/* Respect the kernel parameter if it is set */
+	if (enable_rc6 >= 0) {
+		int mask;
+
+		if (HAS_RC6p(dev_priv))
+			mask = INTEL_RC6_ENABLE | INTEL_RC6p_ENABLE |
+			       INTEL_RC6pp_ENABLE;
+		else
+			mask = INTEL_RC6_ENABLE;
 
-	return info->has_rc6;
+		if ((enable_rc6 & mask) != enable_rc6)
+			DRM_DEBUG_DRIVER("Adjusting RC6 mask to %d "
+					 "(requested %d, valid %d)\n",
+					 enable_rc6 & mask, enable_rc6, mask);
+
+		return enable_rc6 & mask;
+	}
+
+	if (IS_IVYBRIDGE(dev_priv))
+		return (INTEL_RC6_ENABLE | INTEL_RC6p_ENABLE);
+
+	return INTEL_RC6_ENABLE;
 }
 
 static void gen6_init_rps_frequencies(struct drm_i915_private *dev_priv)
@@ -6970,7 +7005,7 @@
 {
 	struct intel_engine_cs *engine;
 	enum intel_engine_id id;
-	u32 rc6_mode;
+	u32 rc6_mode, rc6_mask = 0;
 
 	/* 1a: Software RC state - RC0 */
 	I915_WRITE(GEN6_RC_STATE, 0);
@@ -7031,6 +7066,9 @@
 	I915_WRITE(GEN9_RENDER_PG_IDLE_HYSTERESIS, 250);
 
 	/* 3a: Enable RC6 */
+	if (intel_rc6_enabled() & INTEL_RC6_ENABLE)
+		rc6_mask = GEN6_RC_CTL_RC6_ENABLE;
+	DRM_INFO("RC6 %s\n", onoff(rc6_mask & GEN6_RC_CTL_RC6_ENABLE));
 	I915_WRITE(GEN6_RC6_THRESHOLD, 37500); /* 37.5/125ms per EI */
 
 	/* WaRsUseTimeoutMode:cnl (pre-prod) */
@@ -7040,9 +7078,7 @@
 		rc6_mode = GEN6_RC_CTL_EI_MODE(1);
 
 	I915_WRITE(GEN6_RC_CONTROL,
-		   GEN6_RC_CTL_HW_ENABLE |
-		   GEN6_RC_CTL_RC6_ENABLE |
-		   rc6_mode);
+		   GEN6_RC_CTL_HW_ENABLE | rc6_mode | rc6_mask);
 
 	/*
 	 * 3b: Enable Coarse Power Gating only when RC6 is enabled.
@@ -7051,8 +7087,8 @@
 	if (NEEDS_WaRsDisableCoarsePowerGating(dev_priv))
 		I915_WRITE(GEN9_PG_ENABLE, 0);
 	else
-		I915_WRITE(GEN9_PG_ENABLE,
-			   GEN9_RENDER_PG_ENABLE | GEN9_MEDIA_PG_ENABLE);
+		I915_WRITE(GEN9_PG_ENABLE, (rc6_mask & GEN6_RC_CTL_RC6_ENABLE) ?
+				(GEN9_RENDER_PG_ENABLE | GEN9_MEDIA_PG_ENABLE) : 0);
 
 	intel_uncore_forcewake_put(dev_priv, FORCEWAKE_ALL);
 }
@@ -7061,6 +7097,7 @@
 {
 	struct intel_engine_cs *engine;
 	enum intel_engine_id id;
+	uint32_t rc6_mask = 0;
 
 	/* 1a: Software RC state - RC0 */
 	I915_WRITE(GEN6_RC_STATE, 0);
@@ -7082,11 +7119,13 @@
 	I915_WRITE(GEN6_RC6_THRESHOLD, 625); /* 800us/1.28 for TO */
 
 	/* 3: Enable RC6 */
-
-	I915_WRITE(GEN6_RC_CONTROL,
-		   GEN6_RC_CTL_HW_ENABLE |
-		   GEN7_RC_CTL_TO_MODE |
-		   GEN6_RC_CTL_RC6_ENABLE);
+	if (intel_rc6_enabled() & INTEL_RC6_ENABLE)
+		rc6_mask = GEN6_RC_CTL_RC6_ENABLE;
+	intel_print_rc6_info(dev_priv, rc6_mask);
+
+	I915_WRITE(GEN6_RC_CONTROL, GEN6_RC_CTL_HW_ENABLE |
+			GEN7_RC_CTL_TO_MODE |
+			rc6_mask);
 
 	intel_uncore_forcewake_put(dev_priv, FORCEWAKE_ALL);
 }
@@ -7135,8 +7174,9 @@
 {
 	struct intel_engine_cs *engine;
 	enum intel_engine_id id;
-	u32 rc6vids, rc6_mask;
+	u32 rc6vids, rc6_mask = 0;
 	u32 gtfifodbg;
+	int rc6_mode;
 	int ret;
 
 	I915_WRITE(GEN6_RC_STATE, 0);
@@ -7171,12 +7211,22 @@
 	I915_WRITE(GEN6_RC6p_THRESHOLD, 150000);
 	I915_WRITE(GEN6_RC6pp_THRESHOLD, 64000); /* unused */
 
+	/* Check if we are enabling RC6 */
+	rc6_mode = intel_rc6_enabled();
+	if (rc6_mode & INTEL_RC6_ENABLE)
+		rc6_mask |= GEN6_RC_CTL_RC6_ENABLE;
+
 	/* We don't use those on Haswell */
-	rc6_mask = GEN6_RC_CTL_RC6_ENABLE;
-	if (HAS_RC6p(dev_priv))
-		rc6_mask |= GEN6_RC_CTL_RC6p_ENABLE;
-	if (HAS_RC6pp(dev_priv))
-		rc6_mask |= GEN6_RC_CTL_RC6pp_ENABLE;
+	if (!IS_HASWELL(dev_priv)) {
+		if (rc6_mode & INTEL_RC6p_ENABLE)
+			rc6_mask |= GEN6_RC_CTL_RC6p_ENABLE;
+
+		if (rc6_mode & INTEL_RC6pp_ENABLE)
+			rc6_mask |= GEN6_RC_CTL_RC6pp_ENABLE;
+	}
+
+	intel_print_rc6_info(dev_priv, rc6_mask);
+
 	I915_WRITE(GEN6_RC_CONTROL,
 		   rc6_mask |
 		   GEN6_RC_CTL_EI_MODE(1) |
@@ -7624,7 +7674,7 @@
 {
 	struct intel_engine_cs *engine;
 	enum intel_engine_id id;
-	u32 gtfifodbg, rc6_mode, pcbr;
+	u32 gtfifodbg, rc6_mode = 0, pcbr;
 
 	gtfifodbg = I915_READ(GTFIFODBG) & ~(GT_FIFO_SBDEDICATE_FREE_ENTRY_CHV |
 					     GT_FIFO_FREE_ENTRIES_CHV);
@@ -7665,9 +7715,10 @@
 	pcbr = I915_READ(VLV_PCBR);
 
 	/* 3: Enable RC6 */
-	rc6_mode = 0;
-	if (pcbr >> VLV_PCBR_ADDR_SHIFT)
+	if ((intel_rc6_enabled() & INTEL_RC6_ENABLE) &&
+	    (pcbr >> VLV_PCBR_ADDR_SHIFT))
 		rc6_mode = GEN7_RC_CTL_TO_MODE;
+
 	I915_WRITE(GEN6_RC_CONTROL, rc6_mode);
 
 	intel_uncore_forcewake_put(dev_priv, FORCEWAKE_ALL);
@@ -7719,7 +7770,7 @@
 {
 	struct intel_engine_cs *engine;
 	enum intel_engine_id id;
-	u32 gtfifodbg;
+	u32 gtfifodbg, rc6_mode = 0;
 
 	valleyview_check_pctx(dev_priv);
 
@@ -7752,8 +7803,12 @@
 				      VLV_MEDIA_RC6_COUNT_EN |
 				      VLV_RENDER_RC6_COUNT_EN));
 
-	I915_WRITE(GEN6_RC_CONTROL,
-		   GEN7_RC_CTL_TO_MODE | VLV_RC_CTL_CTX_RST_PARALLEL);
+	if (intel_rc6_enabled() & INTEL_RC6_ENABLE)
+		rc6_mode = GEN7_RC_CTL_TO_MODE | VLV_RC_CTL_CTX_RST_PARALLEL;
+
+	intel_print_rc6_info(dev_priv, rc6_mode);
+
+	I915_WRITE(GEN6_RC_CONTROL, rc6_mode);
 
 	intel_uncore_forcewake_put(dev_priv, FORCEWAKE_ALL);
 }
@@ -8286,7 +8341,7 @@
 	 * RPM depends on RC6 to save restore the GT HW context, so make RC6 a
 	 * requirement.
 	 */
-	if (!sanitize_rc6(dev_priv)) {
+	if (!i915_modparams.enable_rc6) {
 		DRM_INFO("RC6 disabled, disabling runtime PM support\n");
 		pm_runtime_get(&dev_priv->drm.pdev->dev);
 	}
@@ -8339,7 +8394,7 @@
 	if (IS_VALLEYVIEW(dev_priv))
 		valleyview_cleanup_gt_powersave(dev_priv);
 
-	if (!HAS_RC6(dev_priv))
+	if (!i915_modparams.enable_rc6)
 		pm_runtime_put(&dev_priv->drm.pdev->dev);
 }
 
@@ -8509,8 +8564,7 @@
 
 	mutex_lock(&dev_priv->pcu_lock);
 
-	if (HAS_RC6(dev_priv))
-		intel_enable_rc6(dev_priv);
+	intel_enable_rc6(dev_priv);
 	intel_enable_rps(dev_priv);
 	if (HAS_LLC(dev_priv))
 		intel_enable_llc_pstate(dev_priv);
@@ -9830,7 +9884,7 @@
 	unsigned int i;
 	u32 mul, div;
 
-	if (!HAS_RC6(dev_priv))
+	if (!intel_rc6_enabled())
 		return 0;
 
 	/*
diff -ura archlinux-linux.orig/drivers/gpu/drm/i915/intel_uncore.c archlinux-linux.new/drivers/gpu/drm/i915/intel_uncore.c
--- archlinux-linux.orig/drivers/gpu/drm/i915/intel_uncore.c	2018-09-13 07:46:24.118646105 -0600
+++ archlinux-linux.new/drivers/gpu/drm/i915/intel_uncore.c	2018-09-13 07:43:57.447676586 -0600
@@ -594,6 +594,9 @@
 
 void intel_uncore_sanitize(struct drm_i915_private *dev_priv)
 {
+	i915_modparams.enable_rc6 =
+		sanitize_rc6_option(dev_priv, i915_modparams.enable_rc6);
+
 	/* BIOS often leaves RC6 enabled, but disable it for hw init */
 	intel_sanitize_gt_powersave(dev_priv);
 }
